electrum/lnpeer.py

```diff
@@ -1416,6 +1527,13 @@ async def close_channel(self, chan_id: bytes):
     @log_exceptions
     async def on_shutdown(self, chan: Channel, payload):
         their_scriptpubkey = payload['scriptpubkey']
+        their_upfront_scriptpubkey = chan.config[REMOTE].upfront_shutdown_script
+
+        # BOLT-02 check if they use the upfront shutdown script they advertized
+        if their_upfront_scriptpubkey:
+            if not (their_scriptpubkey == their_upfront_scriptpubkey):
+                raise Exception("remote didn't use upfront shutdown script it commited to in channel opening")
```

Comment:
As you have rightly pointed out on IRC, this exception will not be handled anywhere, it will just get logged and discarded. If you want to test for it, I suggest changing `process_message` to run the coroutine in the Peer's TaskGroup instead.

See patch:

```patch
diff --git a/electrum/lnpeer.py b/electrum/lnpeer.py
index 5a7109eea..6550b9c30 100644
--- a/electrum/lnpeer.py
+++ b/electrum/lnpeer.py
@@ -193,7 +193,7 @@ class Peer(Logger):
                 payload['raw'] = message
             execution_result = f(*args)
             if asyncio.iscoroutinefunction(f):
-                asyncio.ensure_future(execution_result)
+                asyncio.ensure_future(self.taskgroup.spawn(execution_result))
 
     def on_error(self, payload):
         self.logger.info(f"remote peer sent error [DO NOT TRUST THIS MESSAGE]: {payload['data'].decode('ascii')}")
```

This should ensure that the exception gets re-raised in `peer.main_loop()`, around line 305, and will result in us disconnecting from the peer.

In the tests, I think `peer.main_loop` is not actually used atm, so the taskgroup is not started; you would need to change this of course. To be able to test for the exception being raised, there are multiple ways you could go - for example, (1) instead of calling `peer.main_loop`, you could start the taskgroup manually in the unit test;
(2) or you could modify `peer.main_loop` by adding a try-except that handles the exception in the unit-test case and then re-raises it.
Re (2), I mean
```
try:
    async with self.taskgroup as group:
        ...
except Exception as e:
    self.main_loop_exiting_with_exc(e)
    raise
```
and `self.main_loop_exiting_with_exc` could be a no-op normally, but e.g. `Peer` could be subclassed in the unit tests and  `main_loop_exiting_with_exc` overridden.
There are probably even more ways to do it though.

-----

Regardless of implementation details, I recommend changing `process_message` to use the peer's taskgroup, as it's not nice not to await these coroutines anywhere. If you do that change, please consider other consequences, for example, the `@log_exceptions` decorator for `on_shutdown` becomes useless and annoying; and take a look at other affected message handlers whose exceptions will now not be ignored anymore; i.e. grep for `async def on_`.

Reply:
nice patch.